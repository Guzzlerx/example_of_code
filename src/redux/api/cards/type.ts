import { IMyCardProductItem } from 'routes/Main/Cards/CardProducts/type';
import { ICardStatusSwitcherProps } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { IMyCardItem } from 'routes/Main/Cards/MyCards/type';
import {
  CARD_NAMES_ENUM,
  CARD_STATUSES_ENUM,
  CARD_TYPES_ENUM,
  CURRENCY_CODES_ENUM,
  PAYMENT_SYSTEMS_ENUM,
} from 'constants/cards/cards';

export interface IMyCardsList {
  cards: IMyCardItem[];
}

export interface IMyCardDetails {
  creditLimit: string;
  principalDebt: string;
  terminationDate: string;
  accountId: string;
  accountNumber: string;
  balance: string;
  cardId: string;
  cardNumber: string;
  clientId: string;
  currencyCode: CURRENCY_CODES_ENUM;
  expirationDate: string;
  holderName: string;
  paymentSystem: PAYMENT_SYSTEMS_ENUM;
  productName: CARD_NAMES_ENUM;
  status: CARD_STATUSES_ENUM;
  transactionLimit: string;
}

export interface ICardProductsList {
  cardProducts: IMyCardProductItem[];
}

export type TUpdateCardStatusRequest = Omit<ICardStatusSwitcherProps, 'refetchHandler'>;
export interface INewCardRequest {
  idProduct?: string;
  holderName?: string;
  cardType?: CARD_TYPES_ENUM;
}
