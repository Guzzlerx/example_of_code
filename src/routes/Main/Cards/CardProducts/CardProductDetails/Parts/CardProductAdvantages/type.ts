import { CARD_NAMES_ENUM } from 'constants/cards/cards';
import { ICardConfigsAdvantages } from 'constants/cards/types';

export interface ICardProductAdvantagesProps {
  cardName: Exclude<CARD_NAMES_ENUM, CARD_NAMES_ENUM.PERSONAL_LOAN>;
  data: ICardConfigsAdvantages[];
}
