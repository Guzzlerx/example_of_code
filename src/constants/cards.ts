export enum CARD_TYPES_ENUM {
  DEBIT = 'DEBIT',
  CREDIT = 'CREDIT',
  DEPOSIT = 'DEPOSIT',
}

export enum CARD_STATUSES_ENUM {
  ACTIVE = 'ACTIVE',
  BLOCKED = 'BLOCKED',
  EXPIRED = 'EXPIRED',
}

export enum PAYMENT_SYSTEMS_ENUM {
  VISA = 'VISA',
  MASTERCARD = 'MASTERCARD',
}

export enum CARD_NAMES_ENUM {
  UNIVERSAL_CARD = 'universal_card',
  SMART_PAYOUT = 'smart_payout',
  PURCHASE_PLUS = 'purchase_plus',
  PERSONAL_LOAN = 'personal_loan',
  INFINITY_CLASSIC = 'infinity_classic',
  INFINITY_GOLD = 'infinity_gold',
  INFINITY_PREMIUM = 'infinity_premium',
}

export enum CURRENCY_CODES_ENUM {
  EUR = 'EUR',
  USD = 'USD',
  KZT = 'KZT',
  GBP = 'GBP',
  CHF = 'CHF',
  PLN = 'PLN',
  RUB = 'RUB',
}

export const MAX_CARDS_TO_SHOW = 3;

export enum CARD_SIZE_ENUM {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  EXTRA_SMALL = 'extraSmall',
}
