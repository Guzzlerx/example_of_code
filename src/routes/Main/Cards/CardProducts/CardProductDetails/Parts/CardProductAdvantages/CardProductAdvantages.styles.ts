import styled from 'styled-components';
import { COLOR_PALETTE } from 'constants/colors/colors';

export const StyledContainer = styled.section`
  display: flex;
  flex-direction: column;
  text-align: start;
`;

export const StyledAdvantagesContainer = styled.div<{ marginBottom?: number }>`
  display: flex;
  gap: 40px;
  margin-bottom: ${({ marginBottom }) => marginBottom}px;
`;

export const StyledAdvantagesWrapper = styled.div<{ width: number; marginTop?: number }>`
  display: flex;
  flex-direction: column;
  gap: 24px;
  width: ${({ width }) => width}px;
  margin-top: ${({ marginTop }) => marginTop}px;
`;

export const StyledText = styled.p`
  font-size: 14px;
  line-height: 20px;
  color: ${COLOR_PALETTE.gray700};

  & strong {
    font-weight: 700;
    font-size: 14px;
  }
`;

export const StyledSpan = styled.span`
  font-weight: 700;
  font-size: 14px;
`;
