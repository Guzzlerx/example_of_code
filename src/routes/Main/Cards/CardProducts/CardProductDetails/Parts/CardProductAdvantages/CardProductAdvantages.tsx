import { useTranslation } from 'react-i18next';
import { AIcon } from 'components/AIcon/AIcon';
import { ATitle } from 'components/ATitle/ATitle';
import {
  StyledAdvantagesContainer,
  StyledAdvantagesWrapper,
  StyledContainer,
  StyledSpan,
  StyledText,
} from './CardProductAdvantages.styles';
import { CardProductAdvantagesItem } from './CardProductAdvantagesItem/CardProductAdvantagesItem';
import { ICardProductAdvantagesProps } from './type';

export const CardProductAdvantages = ({ cardName, data }: ICardProductAdvantagesProps) => {
  const { t } = useTranslation('cards');

  return (
    <StyledContainer>
      <ATitle size={20} weight={700} lineHeight={28} marginBottom={40}>
        {t('card_details.advantages.title')}
      </ATitle>
      <StyledAdvantagesContainer marginBottom={72}>
        {data.map(
          ({ title, descriptionPart1, descriptionPart2, descriptionPart3, icon }, index) => (
            <StyledAdvantagesWrapper width={180} key={index}>
              <AIcon iconTitle={t(title)} variant="cardAdvantages">
                {icon}
              </AIcon>
              <StyledText>
                {t(descriptionPart1)}
                <StyledSpan>{t(descriptionPart2)}</StyledSpan>
                {t(descriptionPart3)}
              </StyledText>
            </StyledAdvantagesWrapper>
          ),
        )}
      </StyledAdvantagesContainer>
      <CardProductAdvantagesItem cardName={cardName} />
    </StyledContainer>
  );
};
