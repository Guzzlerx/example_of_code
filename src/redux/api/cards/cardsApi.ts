import { apiSlice } from 'redux/api/apiSlice';
import { ICardStatusSwitcherResponse } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { CARD_TYPES_ENUM } from 'constants/cards/cards';
import {
  ICardProductsList,
  IMyCardDetails,
  IMyCardsList,
  INewCardRequest,
  TUpdateCardStatusRequest,
} from './type';

export const cardsApi = apiSlice.injectEndpoints({
  endpoints: builder => ({
    getCreditCards: builder.query<IMyCardsList, void>({
      query: () => '/credit-cards',
    }),
    getDebitCards: builder.query<IMyCardsList, void>({
      query: () => '/deposit-cards',
    }),
    getCardDetails: builder.query<IMyCardDetails, { id: string; type: CARD_TYPES_ENUM }>({
      query: ({ id, type }) => `/${type.toLowerCase()}-cards/${id}`,
    }),
    getCardProducts: builder.query<ICardProductsList, void>({
      query: () => '/cards-products',
    }),
    updateCardStatus: builder.mutation<ICardStatusSwitcherResponse, TUpdateCardStatusRequest>({
      query: ({ cardId, currentStatus, cardType }) => ({
        url: `${cardType.toLocaleLowerCase()}-cards/${cardId}`,
        method: 'PATCH',
        body: currentStatus,
      }),
    }),
    getNewCard: builder.mutation<void, INewCardRequest>({
      query: body => ({
        url: '/deposit-cards-orders/new',
        method: 'POST',
        body: body,
      }),
    }),
  }),
});

export const {
  useGetCreditCardsQuery,
  useGetDebitCardsQuery,
  useGetCardProductsQuery,
  useGetCardDetailsQuery,
  useUpdateCardStatusMutation,
  useGetNewCardMutation,
} = cardsApi;
