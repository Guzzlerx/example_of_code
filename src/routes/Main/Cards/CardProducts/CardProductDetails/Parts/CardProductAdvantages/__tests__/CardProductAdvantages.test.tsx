import { RenderResult, render, screen } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { ReactComponent as ThreeDSecure } from 'assets/icons/cardProductBenefits/3d_secure.svg';
import { ReactComponent as AdditionalCashback } from 'assets/icons/cardProductBenefits/additional_cashback.svg';
import { ReactComponent as ATM } from 'assets/icons/cardProductBenefits/atm.svg';
import { ReactComponent as GiftCard } from 'assets/icons/cardProductBenefits/card_giftcard.svg';
import { ReactComponent as MoneyLimit } from 'assets/icons/cardProductBenefits/local_atm.svg';
import { ReactComponent as Insurance } from 'assets/icons/cardProductBenefits/medical_services.svg';
import { ReactComponent as Transfer } from 'assets/icons/cardProductBenefits/transfer.svg';
import { CARD_NAMES_ENUM } from 'constants/cards/cards';
import i18n from 'utils/languages/i18nForTests';
import { CardProductAdvantages } from '../CardProductAdvantages';

describe('CardProductAdvantages', () => {
  let renderResult: RenderResult;

  const data = [
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.description_3`,
      icon: <GiftCard />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.description_3`,
      icon: <MoneyLimit />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.description_3`,
      icon: <ATM />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.description_3`,
      icon: <Transfer />,
    },
  ];

  const rerenderData = [
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.first_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.first_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.first_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.first_advantage.description_3`,
      icon: <ThreeDSecure />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.second_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.second_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.second_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.second_advantage.description_3`,
      icon: <ATM />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.third_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.third_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.third_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.third_advantage.description_3`,
      icon: <Insurance />,
    },
    {
      title: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.fourth_advantage.image_title`,
      descriptionPart1: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.fourth_advantage.description_1`,
      descriptionPart2: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.fourth_advantage.description_2`,
      descriptionPart3: `card_details.advantages.${CARD_NAMES_ENUM.INFINITY_PREMIUM}.icons.fourth_advantage.description_3`,
      icon: <AdditionalCashback />,
    },
  ];

  beforeEach(() => {
    renderResult = render(
      <I18nextProvider i18n={i18n}>
        <CardProductAdvantages cardName={CARD_NAMES_ENUM.UNIVERSAL_CARD} data={data} />
      </I18nextProvider>,
    );
  });

  it('should render correctly', () => {
    expect(renderResult.container).toBeInTheDocument();
  });

  it('snapshot', () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it('icon container has correct width', () => {
    const element = screen.getByTitle(/Free card/i);

    expect(element).toHaveStyle('width: 64px');
  });

  it('render correctly with "infinity premium" card as prop', () => {
    renderResult.rerender(
      <CardProductAdvantages cardName={CARD_NAMES_ENUM.INFINITY_PREMIUM} data={rerenderData} />,
    );

    expect(screen.getByText(/Stay in control of your card security with/i)).toBeInTheDocument();
    expect(screen.getByText(/3-D/i)).toHaveStyle('font-weight: 700');

    expect(
      screen.getByText(/Withdraw money abroad without fees up to per month/i),
    ).toBeInTheDocument();
    expect(screen.getByText(/£ 10,000/)).toHaveStyle('font-weight: 700');

    expect(
      screen.getByText(/Get insurance for your whole family while travelling abroad - \/person/),
    ).toBeInTheDocument();
    expect(screen.getByText(/£ 100,000/)).toHaveStyle('font-weight: 700');

    expect(
      screen.getByText(/Get your cashback - purchases, without any “ifs”/),
    ).toBeInTheDocument();
    expect(screen.getByText(/8% at all/)).toHaveStyle('font-weight: 700');
  });
});
