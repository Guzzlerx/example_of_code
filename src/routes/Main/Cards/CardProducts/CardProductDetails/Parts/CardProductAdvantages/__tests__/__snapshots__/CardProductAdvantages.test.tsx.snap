// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProductAdvantages snapshot 1`] = `
<div>
  <section
    class="sc-dkrFOg eUBOWt"
  >
    <h2
      class="sc-gswNZR eugAJR"
      size="20"
    >
      Advantages you donâ€™t want to lose
    </h2>
    <div
      class="sc-hLBbgP dsXhKp"
    >
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Free card"
        >
          <svg>
            card_giftcard.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          Free card. Free servicing for a whole year
          <span
            class="sc-gKPRtg hokwDS"
          />
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Money limit"
        >
          <svg>
            local_atm.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          Adjustable money limit - change it anytime in your account
          <span
            class="sc-gKPRtg hokwDS"
          />
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="ATM"
        >
          <svg>
            atm.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          No more need to search for the right ATM. Use wherever you want
          <span
            class="sc-gKPRtg hokwDS"
          />
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Transfer"
        >
          <svg>
            transfer.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          Transfer money by phone number without additional money and limits
          <span
            class="sc-gKPRtg hokwDS"
          />
        </p>
      </div>
    </div>
    <div
      class="sc-hLBbgP cmWkZU"
    >
      <svg>
        advantage_graph.svg
      </svg>
      <div
        class="sc-eDvSVe duNAoy"
        width="196"
      >
        <h2
          class="sc-gswNZR hxLdcO"
          size="20"
        >
          8% cashback
        </h2>
        <p
          class="sc-jSUZER gPgFBy"
        >
          On your first GPB 2,000 eligible purchases with two retailers you choose
        </p>
      </div>
      <div
        class="sc-eDvSVe duNAoy"
        width="196"
      >
        <h2
          class="sc-gswNZR hxLdcO"
          size="20"
        >
          3% cashback
        </h2>
        <p
          class="sc-jSUZER gPgFBy"
        >
          On your first GPB 2,000 eligible purchases on your bills
        </p>
      </div>
      <div
        class="sc-eDvSVe duNAoy"
        width="196"
      >
        <h2
          class="sc-gswNZR hxLdcO"
          size="20"
        >
          2% cashback
        </h2>
        <p
          class="sc-jSUZER gPgFBy"
        >
          On all other eligible purchases
        </p>
      </div>
    </div>
  </section>
</div>
`;
